common/src/main/java/com/ericsson/bss/cassandra/ecaudit/common/formatter/LogMessageFormatter.java

```diff
@@ -0,0 +1,164 @@
+ * </ul>
+ * <p>
+ * Conditional formatting of fields is also available, which makes it possible to only log the field value and
+ * its descriptive text if a value exists. The formatting string {@code "user:${USER}, client:${CLIENT_IP}{?, executed from batch:${BATCH_ID}?}"}
+ * will log different things depending on if there is a BATCH_ID or not:
+ * <ul>
+ * <li>"user:Duke, client:1.2.3.4, executed from batch:e501f872-9aab-4f6b-9a52-8ed2f67b1320"</li>
+ * <li>"user:Duke, client:1.2.3.4"</li>
+ * </ul>
+ *
+ * @param <T> The type of long entries the log message formatter should operate on.
+ */
+public class LogMessageFormatter<T>
+{
+    private static final String FIELD_EXP = "\\$\\{(.*?)}"; // Non-greedy matching of field
+    private static final String OPTIONAL_FIELD_EXP = "\\{\\?(.*?)\\$\\{(.*?)}(.*?)\\?}";
+    private static final String COMBINED_FIELDS_EXP = FIELD_EXP + '|' + OPTIONAL_FIELD_EXP;
+    private static final Pattern FIELD_PATTERN = Pattern.compile(COMBINED_FIELDS_EXP);
+
+    private final String logTemplate;
+    private final List<Function<T, String>> configuredFieldFunctions;
+
+    private LogMessageFormatter(Builder<T> builder)
+    {
+        logTemplate = getTemplateFromFormatString(builder);
+        configuredFieldFunctions = getConfiguredFieldFunctions(builder);
+    }
+
+    private static String getTemplateFromFormatString(Builder builder)
+    {
+        return builder.format.replaceAll(builder.escapeExpr, builder.escapeWith)
+                             .replaceAll(COMBINED_FIELDS_EXP, builder.anchor);
+    }
+
+    private static <T> List<Function<T, String>> getConfiguredFieldFunctions(Builder<T> builder)
+    {
+        List<Function<T, String>> fieldFunctions = new ArrayList<>();
+        Matcher matcher = FIELD_PATTERN.matcher(builder.format);
+        while (matcher.find())
+        {
+            String normalField = matcher.group(1);
+            if (normalField != null)
+            {
+                Function<T, Object> fieldFunction = getFieldFunctionOrThrow(normalField, builder);
+                fieldFunctions.add(fieldFunction.andThen(String::valueOf));
+            }
+            else // Optional field
+            {
+                String descriptionLeft = matcher.group(2);
+                String optionalField = matcher.group(3);
+                String descriptionRight = matcher.group(4);
+                Function<T, Object> fieldFunction = getFieldFunctionOrThrow(optionalField, builder);
+                fieldFunctions.add(fieldFunction.andThen(getOptionalDescriptionAndValueIfPresent(descriptionLeft, descriptionRight)));
+            }
+        }
+        return fieldFunctions;
+    }
+
+    static <T> Function<T, Object> getFieldFunctionOrThrow(String field, Builder<T> builder)
+    {
+        if (!builder.availableFields.containsKey(field))
+        {
+            throw new IllegalArgumentException("Unknown log format field: " + field);
```

Comment:
Question - Is it fine to throw a more "generic" exception from this code now?
Since it is moved to common (which does not have a dependency to Cassandra (neither does EcLog)) we can/should no longer throw a ConfigurationException...
So is it fine with a IllegalArgumentException och should we use something else?